trpcultivate_phenotypes/src/TripalCultivateValidator/ValidatorTraits/Headers.php

```diff
@@ -92,132 +92,112 @@ public function setHeaders(array $headers) {
    *  - If an unrecognized header type is requested in the 'types' parameter.
    */
   public function getHeaders(array $types = ['required', 'optional']) {
 
+    // Check that the headers have been set by the setHeaders() method.
+    if (!array_key_exists($this->context_key, $this->context)) {
+      throw new \Exception('Cannot retrieve headers from the context array as one has not been set by setHeaders() method.');
+    }
+
     $valid_types = $this->types;
-    
-    // This use of array_filter will pull out any unrecognized types by 
+
+    // This use of array_filter will pull out any unrecognized types by
     // comparing the 'types' parameter to the types property defined
     // at the top of this class.
-    $invalid_types = array_filter($types, function($type) use($valid_types) { 
-      return !in_array($type, $valid_types); 
+    $invalid_types = array_filter($types, function($type) use($valid_types) {
+      return !in_array($type, $valid_types);
     });
-    
+
     // If any such unrecognized types are detected, throw an exception.
     if (!empty($invalid_types)) {
       $str_invalid_types = implode(', ', $invalid_types);
       $str_valid_types = implode(', ', $valid_types);
+
       throw new \Exception('Cannot retrieve invalid header types: ' . $str_invalid_types . '. Use one of valid types: [' . $str_valid_types . ']');
     }
-   
-    if (array_key_exists($this->context_key, $this->context)) {
-      // At this point, types requested are valid.   
-      $headers = [];
-      $processed_types = 0;
-
-      foreach($types as $type) {
-        if (array_key_exists($type, $this->context[ $this->context_key ])) {
-          $processed_types++;
-          foreach($this->context[ $this->context_key ][ $type ] as $index => $header) {
-            $headers[ $index ] = $header;
-          }
-        }
-      }
-      
-      // All of the header elements with a type matching one of the types 
-      // requested, keyed by the column index and 'header name' as the value.
-      // Only return if all of the types requested were processed (i.e. were
-      // in the headers context array).
-      if ($processed_types == sizeof($types)) {
-        return $headers;
+
+    // Prepare fullset/subset of the importer headers where each header type matches
+    // the header type requested. The final list of headers generated will maintain
+    // the index and order of the header as they appear in headers
+    // set by in the setHeaders method.
+    $requested_headers = [];
+
+    foreach($this->context[ $this->context_key ] as $index => $header) {
+      if (in_array($header['type'], $types)) {
+        $requested_headers[ $index ] = $header['name'];
       }
     }
-    
-    throw new \Exception('Cannot retrieve requested headers as the ' . $this->context_key . ' key was not set in the context array, or one or more types were not set by setHeaders() method.'); 
+
+    return ($requested_headers) ? $requested_headers : [];
```

Comment:
This check is not needed since you set the variable to an empty array on line 193.

```suggestion
    return $requested_headers;
```